kvclient: enable health check in dist_sender

andrewbaptist opened the pull request:
This commit enables the health check in dist_sender and removes the
redundant check and splitting in transport. The cost of checking the
health of connections was significant in tests with high QPS per node.
Using the cached version will minimize this cost.

Epic: none

Release note: None

Commit:
kvclient: cache the sorted list of replicas

Calling OptimizeReplicaOrder is expensive and can show up as a hotspot
in many high QPS workloads. This PR caches the sorted list on a
per-range basis and updates the cache at most once every 3s.

Epic: none

Release note: None

Commit:
kvclient: replace ReplicaSlice with ReplicaSet for routing

This PR replaces ReplicaSlice with ReplicaSet outside of the
DistSender. ReplicaSlice is an internal implementation which
contains information only required for sorting the replicas.
Outside of DistSender the additional sorting information is
unnecessary and unused.

Epic: none
Informs: #112351

Release note: None

Commit:
kvclient: enable health check in dist_sender

This commit enables the health check in dist_sender and removes the
redundant check and splitting in transport. The cost of checking the
health of connections was significant in tests with high QPS per node.
Using the cached version will minimize this cost.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114619)
<!-- Reviewable:end -->

andrewbaptist commented:
Only the last commit needs to be reviewed. The other commits are on separate PRs and will merge before this one. This is a 1-line change to switch the default from false to true and a larger change to remove unnecessary sorting that is now redundant.

The last commit can stand alone, but I wanted to make it clear that it wasn't going to merge prior to the other changes which cache the sorted set of replicas.

arulajmani reviewed (approved):
3rd commit <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 15 of 15 files at r1, 2 of 2 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114619#-:-NlY3ZIX3n4P656ZgZ4a:b-3p4xmd)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist)



<!-- Sent from Reviewable.io -->

Commit:
kvclient: remove retries within sendToReplicas

This change removes a retry loop within sendToReplicas that occured when
we hit a NLHE. We still retry within the outer loop of sendPartialBatch
and removing this also removes the only remaining reordering of the
replicas that are retrieved from the RangeCache.

There was a metric distsender.errors.inleasetransferbackoffs that tracks
the number of retries within this now removed loop that is also removed.

Epic: none

Release note: none

Commit:
kvclient: simplify test

Epic: none

Release note: None

Commit:
kvclient: tighten checks for cache eviction

Previously the cache would evict if the new descriptors was identical to
the old descriptor even if the cached lease was newer. This change
tightens up the conditions for cache eviction.

Epic: none

Release note: None

Commit:
kvclient: remove deprecated metric

This commit removes a metric which counts the number of retries occur
due to NLHE. This loop is being removed in a future commit.

Epic: none

Release note (ops change): Removes the metric
distsender.errors.inleasetransferbackoffs.

Commit:
kvclient: remove unused MoveToFront

This commit removes ths MoveToFront method on the Transport. A major
advantage of this change is that the Transport no longer modifies the
underlying ReplicaSlice. This will allow a future commit to avoid
copying the slice to every transport.

Epic: none

Release note: None
